Update layout expiry behavior

SpencerKlem opened the pull request:
Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

**Fixes issue #545**:

**Description of the changes being introduced by the pull request**:
Adding test cases, has bug
**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -960,6 +960,14 @@ def setUpClass(cls):
+        # dump layout with current date expiry
+        layout = copy.deepcopy(layout_template)
+        layout.signed.expires = (
+            datetime.now()
```
Here, we have an expiry of 2023-07-...

adityasaky replied:
Notice that you don't need to mock datetime to set an expiry. You can set _any_ value you want here and it does not matter.

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -1023,7 +1031,19 @@ def test_verify_passing_double_signed_layout(self):
+    def test_verify_failing_layout_current_datetime(self):
+        """Test pass verification with current date expiry layout."""
+        with patch("""in_toto.verifylib.datetime.datetime""") as mock_date:
+            with patch("""in_toto.verifylib.iso8601""") as mock_date2:
```
I don't think we'll need to patch this.

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -1023,7 +1031,19 @@ def test_verify_passing_double_signed_layout(self):
+                mock_date.now.return_value= datetime(2025, 1, 2, 3, 4, 6)
+                mock_date2.parse_date.return_value = datetime(2025, 1, 2, 3, 4, 6)
+                assert in_toto.verifylib.datetime.datetime.now() == in_toto.verifylib.iso8601.parse_date(datetime.now())
+                layout = Metablock.load(self.layout_current_datetime_expiry_path)
```
I see, so the layout you configure above is not being used.

adityasaky replied:
Let's configure the layout in setUp for consistency. Pick a date, set it as the expiry. Then mock verifylib's datetime here so that the expiry check uses the timestamp you picked.

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -1023,7 +1031,19 @@ def test_verify_passing_double_signed_layout(self):
+                layout = Metablock.load(self.layout_current_datetime_expiry_path)
+                layout_key_dict = import_publickeys_from_file([self.alice_path])
+                with self.assertRaises(LayoutExpiredError):
+                    in_toto_verify(layout, layout_key_dict)
```
```
-> if expire_datetime <= datetime.datetime.now(tz.tzutc()):
(Pdb) expire_datetime
datetime.datetime(2025, 1, 2, 3, 4, 6)
(Pdb) datetime.datetime.now()
datetime.datetime(2025, 1, 2, 3, 4, 6)
(Pdb)
```

A breakpoint in verifylib indicates this works fine **logically**, and the test indeed passes. However, I think the setup can be better.

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -1081,7 +1101,7 @@ def test_verify_failing_inspection_rules(self):
             in_toto_verify(layout, layout_key_dict)
 
-    def test_verify_layout_signatures_fail_with_no_keys(self):
+    def test_signatures_fail_with_no_keys(self):
```
what's this change for, sorry?

adityasaky reviewed (changes requested):
I think logically this coming along. Note that you don't have to close this PR to push more changes, you can just push to the same branch. :)

SpencerKlem commented:
Current issue: verifylib.py is having trouble comparing the two mocked dates as one is timezone aware and the other date is not.

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -1087,7 +1106,7 @@ def test_verify_layout_signatures_fail_with_no_keys(self):
             in_toto_verify(layout_metablock, {})
 
-    def test_verify_layout_signatures_fail_with_malformed_signature(self):
+    def test_signatures_fail_with_no_keys(self):
```
bump on this change, I'm not sure why this test name was updated?

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -1023,7 +1032,17 @@ def test_verify_passing_double_signed_layout(self):
+        """Test pass verification with current date expiry layout."""
+        with patch("""in_toto.verifylib.datetime""") as mock_date:
+            mock_date.datetime.now.return_value= datetime(2025, 1, 2, 3, 4, 6).replace(tzinfo=pytz.utc)
+            assert in_toto.verifylib.datetime.datetime.now() == datetime(2025, 1, 2, 3, 4, 6).replace(tzinfo=pytz.utc)
```
let's get rid of this assertion

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -960,6 +961,14 @@ def setUpClass(cls):
+        # dump layout with current date expiry
+        layout = copy.deepcopy(layout_template)
+        layout.signed.expires = (
+            datetime(2025, 1, 2, 3, 4, 6).replace(tzinfo=pytz.utc)
```
can't tzinfo be passed in to datetime as a named parameter? :)

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -29,6 +29,7 @@
 import shutil
 import tempfile
 import unittest
+import pytz
```
Instead of using another dependency I suggest using `datetime.timezone.utc` or `dateutil.tz.tzutc`. The latter is also a third-party dependency but we already use it for `relativedelta`.

adityasaky replied:
Also the CI's likely to fail as this isn't installed.

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -1023,7 +1032,17 @@ def test_verify_passing_double_signed_layout(self):
+            assert in_toto.verifylib.datetime.datetime.now() == datetime(2025, 1, 2, 3, 4, 6).replace(tzinfo=pytz.utc)
+            layout = Metablock.load(self.layout_single_signed_path)
+            layout_key_dict = import_publickeys_from_file([self.alice_path])
+            with self.assertRaises(LayoutExpiredError):
```
Neat, I think we can also add another where the time is off by a second and the expiry check passes.

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -82,7 +81,7 @@
     verify_threshold_constraints,
 )
-from tests.common import GPGKeysMixin, TmpDirMixin
+from common import GPGKeysMixin, TmpDirMixin
```
This will break the import, no?

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -1032,17 +1040,23 @@ def test_verify_passing_double_signed_layout(self):
-            assert in_toto.verifylib.datetime.datetime.now() == datetime(2025, 1, 2, 3, 4, 6).replace(tzinfo=pytz.utc)
+            mock_date.datetime.now.return_value= datetime(2025, 1, 2, 3, 4, 5, tzinfo=timezone.utc)
+            layout = Metablock.load(self.layout_single_signed_path)
+            layout_key_dict = import_publickeys_from_file([self.alice_path])
```
There should be a call to in_toto_verify here that passes.

SpencerKlem committed:
Update layout expiry

Previously, the layout was considered to be expired if the
current time was after the timestamp in the layout. Now, the
layout is considered to be expired from the time specified
in the layout onwards.

Signed-off-by: Spencer Klem <macandtony43@gmail.com>

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -960,6 +962,22 @@ def setUpClass(cls):
+        layout.dump(cls.layout_near_current_datetime_expiry_path)
+
+        # dump layout with current date expiry
+        layout = copy.deepcopy(layout_template)
```
I think this layout is identical to the one above?

adityasaky replied:
If yes, let's reuse the same layout.

adityasaky commented on tests/test_verifylib.py:
```diff
@@ -1024,6 +1042,23 @@ def test_verify_passing_double_signed_layout(self):
+    def test_verify_failing_layout_current_datetime(self):
+        """Test fail verification with current date expiry layout."""
+        with patch("""in_toto.verifylib.datetime""") as mock_date:
+            mock_date.datetime.now.return_value= datetime(2025, 1, 2, 3, 4, 6, tzinfo=timezone.utc)
```
I think this needs to be 4 instead of 6 as you're changing the current time and not the expire time. Remember that the comparison is `expires <= current` so for that condition to be false, the current time must be at least a second before the expire time.

lukpueh reviewed (approved):
💯

adityasaky commented:
Great work, @SpencerKlem!
